### PR TITLE
[Non-Record] MLP3x + WD0.04 + OrthoInit + Sliding Eval — 1.4536 BPB

### PR DESCRIPTION
**Pull request by AymanMahfuz27**
## Summary

Non-record local submission documenting **38+ controlled experiments** on RTX 3060, systematically identifying techniques that improve post-quantization BPB individually and in combination.

**Final int8+zlib roundtrip: 1.4536 BPB** (local, 1000 steps × 65K batch). Projected **~1.18–1.21 BPB at H100 scale** (13,780 steps × 524K batch).

## What Changed from Baseline

| Technique | Config | Impact | Mechanism |
|-----------|--------|--------|-----------|
| **MLP width ×3** | `MLP_MULT=3` | +0.04 BPB | Widens FFN hidden dim, adds ~3.5M params. Every top entry uses this. |
| **Decoupled weight decay** | `WEIGHT_DECAY=0.04` | +0.007 BPB, quant gap 0.004→0.001 | `p.mul_(1 - lr*wd)` in Muon. Regularizes magnitudes → quantization-friendly. |
| **Orthogonal init** | `ORTHO_INIT=1` | +0.003 BPB | `nn.init.orthogonal_()` for 2D CastedLinear weights. |
| **Sliding window eval** | `EVAL_WINDOW_STRIDE=64` | +0.035 BPB (free) | Overlapping 1024-token windows, mask context with `ignore_index=-100`. |

## Additional Implemented Features (not used in this run, available via env vars)

- **Int6 quantization** (`QUANT_BITS=6`): Clamp to [-31, 31] for ~25% smaller artifacts
- **QAT** (`QAT=1`): STE fake-quant in CastedLinear forward pass
- **Bigram hash embeddings** (`BIGRAM_HASH_SIZE=N`): Hash-based bigram context, zero-init residual
- **zstd-22 compression** (`COMPRESSOR=zstd`): ~40% smaller than zlib-9
- **SWA**: Implemented but **proven quantization-hostile** (0.12 BPB degradation in 4 experiments)

## Key Negative Results

| Dead End | Evidence | Lesson |
|----------|----------|--------|
| SWA + post-quant | 4 experiments, ~0.12 BPB gap | EMA weights cluster tightly, int8 can't represent them |
| Layer sharing | 5 experiments, no improvement | Weight tying hurts more than depth helps at this scale |
| SmearGate (dim=12) | Catastrophic 2.5 BPB | 12-dim bottleneck cripples attention information flow |
| LOGIT_SOFTCAP=50 | Worse than default 30 | Keep at 30 |

## Metrics

| Metric | Value |
|--------|-------|
| Pre-quant val BPB | 1.4522 |
| **Post-quant val BPB** | **1.4536** |
| Quantization gap | 0.0014 (excellent) |
| Model params | 21,778,504 |
| Artifact size (int8+zlib) | 14,595,701 bytes |
| Code size | 64,829 bytes |
| Total | 14,660,530 bytes (< 16 MB) |

## H100 Plan

With compute access, the planned experiment sequence:
1. Reproduce baseline (verify 1.2244 BPB)
2. This config at H100 scale → projected ~1.18–1.21 BPB
3. 10L deeper model (10L fits comfortably under 16 MB with zstd)
4. 11L deeper model (best pre-quant locally, untested without SWA)
5. BigramHash if headroom exists
6. Int6+QAT fallback for deeper models
7. Warmdown tuning (1200 default tuned for 20K steps, ~1000 better for 13K)

## Command

```bash
MLP_MULT=3 WEIGHT_DECAY=0.04 ORTHO_INIT=1 \
EVAL_SLIDING_WINDOW=1 EVAL_WINDOW_STRIDE=64 \
torchrun --standalone --nproc_per_node=8 train_gpt.py
```

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — [Non-Record] MLP3x + WD0.04 + OrthoInit + Sliding Eval — 1.4536 BPB

**BPB:** 1.4536 | **Compliance:** LOOKS CLEAN — pure-neural submission, no TTT/SLOT/n-gram-cache

**What I found in the code** (head SHA `a5bad1c292b2`, file `records/track_non_record_16mb/2026-03-22_MLP3x_WD04_OrthoInit_SlidingEval/train_gpt.py`):

Static code review found no TTT adaptation function, no SLOT optimization loop, no n-gram-cache class, and no pre-quant val-token fine-tune. The eval path uses the standard sliding-window stride-64 pattern. The submission is a pure-neural architecture iteration on the standard SP1024/SP4096/SP8192 baseline.

**CPU smoke test (CT2038 proteus-engine, 2026-04-11):** import OK in 0.05s, dim=512, layers=9, vocab=1024, code=65091 B, SMOKE_TEST_PASS

**Verdict:** LOOKS CLEAN.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). No compliance flags from the classification pass — this looks like a clean pure-neural iteration on the standard baseline.

*Auto-classification caveat:* this review was drafted by the AST-based classifier. If there's a non-standard eval mechanism (logit postprocessing, hedge mixing, etc.) that I missed because it's factored into a helper file or a non-standard function name, please flag it and I'll re-run the audit manually.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). CPU smoke test (CT2038 proteus-engine, 2026-04-11): import OK in 0.05s, dim=512, layers=9, vocab=1024, code=65091 B, SMOKE_TEST_PASS. Classification via deterministic AST-based `classify_prs.py` (pattern bank derived from ~65 manually-reviewed PRs earlier in the 2026-04-11 sweep). This review was auto-drafted from a template and spot-checked before posting — if the template misread your code, please call it out so I can iterate the classifier.*
